src/client.rs

```diff
@@ -0,0 +1,162 @@
+            let ciphertext_res =
+                cipher.encrypt(&nonce, &dummy_data, Option::None);
+
+            if ciphertext_res.is_err() {
+                return Err(Error::new(
+                    ErrorKind::Interrupted,
+                    Result::unwrap_err(ciphertext_res).to_string(),
+                ));
+            }
+
+            let encrypted_dummy =
+                [nonce.as_bytes(), ciphertext_res.unwrap().as_slice()].concat();
+
+            // FIXME- uniform generation for now. Is fine but dummies' paths do
+            // not need to be generated at random.
+            let path = self.csprng.gen_range(0..nb_leaves);
+
+            dummies.push(DataItem::new(encrypted_dummy, path as u16));
+        }
+
+        Ok(dummies)
+    }
+
+    pub fn encrypt_blocks(
+        &mut self,
+        blocks: &mut Vec<DataItem>,
+        block_ids: Vec<usize>,
+        max_path: usize,
+    ) -> Result<(), CryptoCoreError> {
+        self.gen_key();
+        let cipher = Aes256Gcm::new(&self.key);
+
+        let mut i = 0;
+        while i < blocks.len() {
+            // Generate new nonce for encryption.
+            let nonce = Nonce::new(&mut self.csprng);
+
+            let ciphertext_res =
+                cipher.encrypt(&nonce, blocks[i].data(), Option::None);
+            if ciphertext_res.is_err() {
+                return Err(Result::unwrap_err(ciphertext_res));
+            }
+
+            // Change element data to plaintext.
+            blocks[i].set_data(
+                [nonce.as_bytes(), ciphertext_res.unwrap().as_slice()].concat(),
+            );
+
+            /*
+             * If the block is among the ones to change, change its path by
+             * sampling a random uniform distribution.
+             */
+            if block_ids.contains(&i) {
+                blocks[i].set_path(self.csprng.gen_range(0..max_path as u16))
+            }
+
+            i += 1;
+        }
+
+        Ok(())
+    }
+
+    pub fn decrypt_blocks(
+        &self,
+        blocks: &mut Vec<DataItem>,
+    ) -> Result<(), CryptoCoreError> {
+        let cipher = Aes256Gcm::new(&self.key);
+
+        let mut i = 0;
+        while i < blocks.len() {
+            // Edge-case where dummies left cells uninitialized.
+            if blocks[i].data().is_empty() {
+                i += 1;
+                continue;
+            }
+
+            let nonce_res = Nonce::try_from_slice(
+                &blocks[i].data()[..Aes256Gcm::NONCE_LENGTH],
+            );
+            if nonce_res.is_err() {
+                return Err(Result::unwrap_err(nonce_res));
+            }
+
+            let nonce = nonce_res.unwrap();
+            let plaintext_res = cipher.decrypt(
+                &nonce,
+                &blocks[i].data()[Aes256Gcm::NONCE_LENGTH..],
+                Option::None,
+            );
+
+            if plaintext_res.is_err() {
+                return Err(Result::unwrap_err(plaintext_res));
+            }
```

Comment:
I am pretty sure there is a function in `crypto_core` that performs exactly this.

Reply:
I checked and this is not true anymore for the latest release of CryptoCore...